warehouse/forklift/legacy.py

```diff
@@ -362,6 +366,92 @@ def _is_duplicate_file(db_session, filename, hashes):
+    artifact. Attestations are only allowed when uploading via a Trusted
+    Publisher, because a Trusted Publisher provides the identity that will be
+    used to verify the attestations.
+    Currently, only GitHub Actions Trusted Publishers are supported, and
+    attestations are discarded after verification.
+    """
+
+    # Check that if the file was uploaded with attestations, verification
+    # passes
+    if "attestations" in request.POST:
+        publisher = request.oidc_publisher
+        if not publisher or not publisher.publisher_name == "GitHub":
+            raise _exc_with_message(
+                HTTPBadRequest,
+                "Attestations are currently only supported when using Trusted "
+                "Publishing with GitHub Actions.",
+            )
+        try:
+            attestations = TypeAdapter(list[Attestation]).validate_json(
+                request.POST["attestations"]
+            )
+        except ValidationError as e:
+            # Log invalid (malformed) attestation upload
+            metrics.increment("warehouse.upload.attestations.malformed")
+            raise _exc_with_message(
+                HTTPBadRequest,
+                f"Error while decoding the included attestation: {e}",
+            )
+
+        if len(attestations) > 1:
+            metrics.increment(
+                "warehouse.upload.attestations." "failed_multiple_attestations"
+            )
+            raise _exc_with_message(
+                HTTPBadRequest,
+                "Only a single attestation per-file is supported at the moment.",
+            )
+
+        verification_policy = publisher.publisher_verification_policy(
+            request.oidc_claims
+        )
+        for attestation_model in attestations:
+            try:
+                # For now, attestations are not stored, just verified
+                predicate_type, _ = attestation_model.verify(
+                    Verifier.production(),
+                    verification_policy,
+                    artifact_path,
+                )
+            except VerificationError as e:
+                # Log invalid (failed verification) attestation upload
+                metrics.increment("warehouse.upload.attestations.failed_verify")
+                raise _exc_with_message(
+                    HTTPBadRequest,
+                    f"Could not verify the uploaded artifact using the included "
+                    f"attestation: {e}",
+                )
+            except Exception as e:
+                sentry_sdk.capture_message(
+                    f"Unexpected error while verifying attestation: {e}"
+                )
+                raise _exc_with_message(
+                    HTTPBadRequest,
+                    f"Unknown error while trying to verify included "
+                    f"attestations: {e}",
+                )
+
+            if predicate_type != "https://docs.pypi.org/attestations/publish/v1":
```

Comment:
Nitpick: should probably constant-ify this predicate type somewhere, or pull it directly from `pypi-attestations` if possible.

Reply:
https://github.com/trailofbits/pypi-attestations/pull/29 will add an appropriate constant.